lom/src/engine/context.go

```diff
@@ -0,0 +1,394 @@
+ * Publish API is generally pretty quick, a value of 10 is more than enough.
+ */
+const HEARTBEAT_CH_SIZE = 10
+
+/* Cache of all registrations & heartbeats from clients. */
+type ClientRegistrations_t struct {
+    activeActions   map[string]*ActiveActionInfo_t
+    activeClients   map[string]*ActiveClientInfo_t
+    heartbeatCh     chan string     /* Writes name of Action sending HB */
+}
+
+/* Initialized object; But not exported */
+var clientRegistrations *ClientRegistrations_t = nil
+
+
+func GetRegistrations() *ClientRegistrations_t {
+    if clientRegistrations == nil {
+        InitRegistrations()
+    }
+    return clientRegistrations
+}
+
+
+func InitRegistrations() *ClientRegistrations_t {
+    clientRegistrations = &ClientRegistrations_t{
+                make(map[string]*ActiveActionInfo_t),
+                make(map[string]*ActiveClientInfo_t),
+                make(chan string, HEARTBEAT_CH_SIZE),
+            }
+    go clientRegistrations.PublishHeartbeats()
+    return clientRegistrations
+}
+
+func (p *ClientRegistrations_t) RegisterClient(name string) error {
+    if len(name) == 0 {
+        return LogError("Expect non empty name")
+    }
+    if _, ok := p.activeClients[name]; ok {
+        LogInfo("%s: Duplicate client registration; De & re-register", name)
+        p.DeregisterClient(name)
+    }
+    cl := &ActiveClientInfo_t {
+        ClientName: name,
+        Actions: make(map[string]struct{}),
+        pendingWriteRequests: make(chan *ServerRequestData, CHAN_REQ_SIZE),
+        pendingReadRequests: make(chan *LoMRequestInt, CHAN_REQ_SIZE),
+        abortCh: make(chan interface{}, 2) }
+                
+    p.activeClients[name] = cl
+    go cl.ProcessSendRequests()
+    return nil
+}
+
+func (p *ClientRegistrations_t) RegisterAction(actionInfo *ActiveActionInfo_t) error {
+    if actionInfo == nil {
+        return LogError("Expect non nil ActiveActionInfo_t")
+    }
+    cl, ok := p.activeClients[actionInfo.Client]
+    if !ok {
+        return LogError("%s: Missing client registration", actionInfo.Client)
+    } else if r, ok1 := p.activeActions[actionInfo.Action]; ok1 {
+        LogInfo("%s/%s: Duplicate action registration (%s); De/re-register,",
+                actionInfo.Client, actionInfo.Action, r.Client)
+        p.DeregisterAction("", actionInfo.Action)
+    }
+    if cfg, err := GetConfigMgr().GetActionConfig(actionInfo.Action); err != nil {
+        return LogError("%s: Missing action config", actionInfo.Action)
+    } else if cfg.Disable {
+        return LogError("%s: is disabled in config", actionInfo.Action)
+    } else {
+        if actionInfo.Timeout == 0 {
+            actionInfo.Timeout = cfg.Timeout
+        }
+        cl.Actions[actionInfo.Action] = struct{}{}
+
+        /* Make a copy and save */
+        info := &ActiveActionInfo_t{}
+        *info = *actionInfo
+        p.activeActions[actionInfo.Action] = info
+        GetSeqHandler().RaiseRequest(actionInfo.Action)
```

Comment:
if RaiseRequest() fails due to channel full, then there is no retry ? Client may not be registered right?

Reply:
No. This is nothing to do with client registration.
RaiseRequest is for sending ActionRequest to an action, if it is first in sequence. 
If fails due to chan full, it implies that client has not been reading for a while. This is a buggy state, which is logged as error.